compiler/src/dotty/tools/dotc/transform/init/Errors.scala

```diff
@@ -0,0 +1,116 @@
+        val pos = tree.sourcePos
+        val line = "[ " + pos.source.file.name + ":" + (pos.line + 1) + " ]"
+        if (last != line)
+          sb.append(
+            if (pos.source.exists)
+              i"${ " " * indentCount }-> ${pos.lineContent.trim}\t$line\n"
+            else
+              i"${tree.show}\n"
+          )
+        last = line
+      }
+      sb.toString
+    }
+
+    /** Flatten UnsafePromotion errors
+     */
+    def flatten: Errors = this match {
+      case unsafe: UnsafePromotion => unsafe.errors.flatMap(_.flatten)
+      case _ => Set(this)
+    }
+  }
+
+  /** Access non-initialized field */
+  case class AccessNonInit(field: Symbol, trace: Vector[Tree]) extends Error {
+    def source: Tree = trace.last
+    def show(implicit ctx: Context): String =
+      "Access non-initialized field " + field.show + "."
+
+    override def report(implicit ctx: Context): Unit = ctx.error(show + stacktrace, field.sourcePos)
+  }
+
+  /** Promote `this` under initialization to fully-initialized */
+  case class PromoteThis(pot: ThisRef, source: Tree, trace: Vector[Tree]) extends Error {
+    def show(implicit ctx: Context): String = "Promote `this` to be initialized while it is not."
+  }
+
+  /** Promote `this` under initialization to fully-initialized */
+  case class PromoteWarm(pot: Warm, source: Tree, trace: Vector[Tree]) extends Error {
+    def show(implicit ctx: Context): String =
+      "Promoting the value under initialization to be initialized."
+  }
+
+  /** Promote a cold value under initialization to fully-initialized */
+  case class PromoteCold(source: Tree, trace: Vector[Tree]) extends Error {
+    def show(implicit ctx: Context): String =
+      "Promoting the value " + source.show + " to be initialized while it is under initialization" + "."
+  }
+
+  case class AccessCold(field: Symbol, source: Tree, trace: Vector[Tree]) extends Error {
+    def show(implicit ctx: Context): String =
+      "Access field " + source.show + " on a value under unknown initialization status" + "."
+  }
+
+  case class CallCold(meth: Symbol, source: Tree, trace: Vector[Tree]) extends Error {
+    def show(implicit ctx: Context): String =
+      "Call method " + source.show + " on a value under unknown initialization status" + "."
+  }
+
+  case class CallUnknown(meth: Symbol, source: Tree, trace: Vector[Tree]) extends Error {
+    def show(implicit ctx: Context): String =
+      "Calling the external method " + meth.show + " may cause initialization errors" + "."
+  }
+
+  /** Promote a value under initialization to fully-initialized */
+  case class UnsafePromotion(pot: Potential, source: Tree, trace: Vector[Tree], errors: Errors) extends Error {
+    assert(errors.nonEmpty)
+
+    override def report(implicit ctx: Context): Unit = ctx.warning(show, source.sourcePos)
+
+    def show(implicit ctx: Context): String = {
+      var index = 0
+      "Promoting the value to initialized is unsafe.\n" + stacktrace +
```

Comment:
Here I also had trouble understanding the error message. The same problem as with `PromoteThis`, the wording could be better.